Add hub2graph script

vmarkovtsev opened the pull request:
The problem which this script solves is to convert a Hub module
to the self-contained GraphDef protobuf and also optionally to
Tensorboard logs. E.g.
```
python3 hub2graph.py -m https://tfhub.dev/google/imagenet/mobilenet_v2_140_224/classification/1 -o mobilenetv2.pb -t tblogs
```
Thus it becomes possible to fetch a Hub module and run inference in Go for example, which
does not have the Hub support yet. Besides, it is handly for quick visualization of a module.

Signed-off-by: Vadim Markovtsev <vadim@sourced.tech>

andresusanopinto commented:
Hi @vmarkovtsev, thanks for the pull request.

I have a few concerns regarding this commit which I am still trying to wrap my head around.

- A particular issue with GraphDefs with frozen variables is that they become really large and many tools have issues processing them. This may work with the simplest modules but it will lead to problems with others.

- Also I am not familiar with level of go support. Is there any sort of support for SavedModels with variables? If there is then working in a tool for saved model or a more unified story there could leverage the tools. 

- I wonder if Tensorboard should allow to load a MetaGraphDef, this would also mean that less manual steps are needed to visualize a Module graph.

vmarkovtsev commented:
Hi @andresusanopinto 

Regarding the size, I don't get you. The structure of the graph is the same, we only change some variable types to constants. Examples: mobilenetv2 takes 26MB in Hub, 24MB in GraphDef; resnet101 175MB in Hub, 171MB in GraphDef. Both work perfectly. I don't see reasons why something does not work because all these hipster MetaGraphDefs are the same good old graphs in the core. Anyway, you know better so I admit it is possible. My use cases are clearly different from running really large graphs, and I can add a disclaimer text in the header. Source: https://www.tensorflow.org/mobile/prepare_models

There is a SavedModel support in Go. Are Hub cache subdirs SavedModels?
However, this has nothing to do with this PR as Go application is just an example, not the goal. The goal is to write the GraphDef.

Regarding Tensorboard, I don't think so unfortunately. It requires `tf.summary.FileWriter` or `tf.train.Saver` which write `events`-style files.

Overall, I must say that the current Tensorflow model formats are a mess. We've got:

- GraphDef
- MetaGraph
- SavedModel
- Checkpoint
- Hub

And all work :)

vmarkovtsev commented:
Update: yes, the following code indeed works:

```
import tensorflow as tf

with tf.Session(graph=tf.Graph()) as sess:
    tf.saved_model.loader.load(sess, [tf.saved_model.tag_constants.TRAINING], "/tmp/tfhub_modules/62e726661d2cbcfce3d1acb031ce5224587f5a18")
```

vmarkovtsev commented:
Update 2: SavedModel support in Go is pretty basic, it does not support these nice "inputs" and "outputs" from MetaGraph and we have to specify the operation/tensor names by hand. After all, it is required to dig and find them out (the PR-ed script does).

andresusanopinto commented:
- There is a 2Gb limit on proto serialization, but before things crash at that level we have seen performance degradation in having a large proto (containing the checkpoint) being copied around in some libraries. But ok, I see a point in having an example on how to freeze a module, do you think having this in the examples directory would suit?

andresusanopinto commented on tensorflow_hub/hub2graph.py:
```diff
@@ -0,0 +1,143 @@
+        # everything happens inside the single default graph
+        log.info("Importing %s", args.hub)
+        module = tfhub.Module(args.hub)
+        signature_def = module._impl._meta_graph.signature_def.get(
```
Could you create tf.placeholders with known names and feed them to module.__call__() instead of depending on internals?

vmarkovtsev replied:
I had `__call__` in the initial version of the script  it there are problems with it. It creates a new subgraph and links variables from the initial graph. It takes considerable time (up to 10 seconds). Besides, it gives a nasty long name to the subgraph (totally sane for the intended use case) which is hard to change. It also inserts placeholders before placeholders - working solution, but not the cleanest for Tensorboard.

vmarkovtsev commented:
@andresusanopinto Done:

1. Moved to examples. Ready to add docs if required.
2. Added comments regarding the graph size and using the protected members

vmarkovtsev commented:
Update 3: loading SavedModel in Go fails with
```
FeedInputs: unable to find feed output
```
Alas!

vmarkovtsev commented:
@andresusanopinto ping

andresusanopinto commented:
After spending more time thinking, we decided that we do not want to have examples that depend on implementation details. So you would need to maintain this as a fork on your side if you want to.

For an example on how to freeze a graph not using the internal APIs check [tfjs converter from Hub Module](https://github.com/tensorflow/tfjs-converter/blob/master/python/tensorflowjs/converters/tf_saved_model_conversion.py#L343)  via freezing a new graph with a Module.
